### DISCUSSION TITLE
Ревью ресурса ServiceRequest в формате fsh

### DISCUSSION THREAD

**Discussion by avzviagin** [fhir-ru/core#228] [General]
Для обмена с POCT (Point Of Care Testing) подготовил пример ресурса ServiceRequest в формате fsh (ServiceRequest.txt).
Предполагаю начать с корневого профиля, затем уточнять его в профилях СЭМД и POCT.
В онлайн генераторе (https://fshschool.org/FSHOnline/#/) ошибок генерации по этому примеру нет.
Не удалось сослаться на корневой ресурс Patient: добавлял в начале описание корневого профиля пациента, потом указывал на него ссылку (ServiceRequestPatient.txt). Этот пример дает ошибку в генераторе: The type "CorePatient" does not match any of the allowed types: Reference(http://hl7.org/fhir/StructureDefinition/Patient | http://hl7.org/fhir/StructureDefinition/Group | http://hl7.org/fhir/StructureDefinition/Location | http://hl7.org/fhir/StructureDefinition/Device)
  Line: 21.
Прошу провести ревью примеров.
[ServiceRequest.txt](https://github.com/fhir-ru/core/files/11598697/ServiceRequest.txt)
[ServiceRequestPatient.txt](https://github.com/fhir-ru/core/files/11598699/ServiceRequestPatient.txt)

**Comment by eugenekogan-spb** [comment]
Я предположил, что проблема у ServiceRequest, в котором указан subject типа core-Patient, возникает из-за того, что профиль core-Patient  не включен в пакет. Я включил его в его текущем состоянии (этот сокращенный профиль сейчас опубликован на  http://fhir.ru/core и сам по себе компилируется без ошибок), но получил ту же ошибку.

**Comment by Vrodvi** [comment]
Проблемные места указанного профиля:

1. Ссылка на пациента - это тип данных Reference, поэтому и указывать нужно не просто название профиля пациента, а указывать его с типом данных Reference (см. подробнее про профилирование ссылок тут - https://build.fhir.org/ig/HL7/fhir-shorthand/reference.html#references)
1.1. Также есть проблема с тем, что в корневом профиле не должно быть слишком жёстких ограничений, а тут происходит попытка ограничить исходную ссылку на два типа ресурсов до ссылки только на пациента
исходно описание такое - `Reference(Patient | Group)`. Следует добавить недостающий корневой профиль и на Group и учесть его в этой ссылке.
1.2. Тип данных у этого атрибута - уже ссылка. Поэтому дополнительный комментарий о том, что это ссылка - лишний.

Т.е. вместо 
```
* subject 1..1
* subject ^short = "Субъект запроса"
* subject only CorePatient

* subject.reference 1..1
* subject.reference ^short = "Ссылка на субъект запроса"
```
должно быть
```
* subject 1..1
* subject ^short = "Субъект запроса"
* subject only Reference(CorePatient or CoreGroup)

//* subject.reference 1..1
//* subject.reference ^short = "Ссылка на субъект запроса"
```
также не забыть обозначить (и описать) все упомянутые профили, а именно:
```
Profile:     CorePatient
Id:          core-patient
Parent:      Patient
Title:       "Пациент (Ru Core)"
Description: "Пациент"

Profile:     CoreGroup
Id:          core-group
Parent:      Patient
Title:       "TODO (Ru Core)"
Description: "TODO"
```


2. В блоке описаний для атрибута `code` слишком жёстко ограничено предназначение. Это корневой профиль и он должен быть максимально открытым и приближенным к исходному ресурсу.
В данном случае общее описание `What's being requested/ordered` заменено на узкоспециализированные описания - про услуги и профили. Эти специфические описания должны быть на следующем уровне профилей - на уровне профилей под конкретную задачу. Тоже самое касается срезов для кодов: указанные срезы - это уже специфика конкретной области.
Описания необходимо изменить на более общие.
Срезы необходимо вынести в следующий уровень профилирования.

3. По самим срезам также есть ошибки.
Если срез определяется для элемента `.coding`, то в пути discriminator.path он уже не должен указываться. В нём нужно указать какой-то вложенный атрибут по которому будет происходить нарезка. 
вместо:
```
discriminator.path = "coding.system" 
```
должно быть:
```
discriminator.path = "system" 
```

4. При указании справочников (CodeSystem) в срезах использован формат URL не по утверждённому регламенту (см. регламент тут - http://fhir-ru.zendoc.me/support.fhir-guide.conventions, конкретно неправильно указан тип ресурса - должен быть `cs` (CodeSystem), а использован `system` (который предназначен для ресурса NamingSystem, который не применяется для непосредственно кодирования данных ))
вместо:
```
* code.coding[НМУ].system = "http://fhir.ru/core/systems/nmu"
* code.coding[ФСЛИ].system = "http://fhir.ru/core/systems/fsliprofile"  
```
должно быть:
```
* code.coding[НМУ].system = "http://fhir.ru/core/cs/core-cs-nsi-nmu"
* code.coding[ФСЛИ].system = "http://fhir.ru/core/cs/core-cs-nsi-fsliprofile"  
```
Плюс все использованные справочники и наборы значений должны быть также описаны тут - http://fhir-ru.zendoc.me/core.codesystems, а указанных справочников там нет.

4.1. @eugenekogan-spb Тут ещё небольшая проблема в документации на Zendoc - раздел [Идентификаторы Coding Systems](http://fhir-ru.zendoc.me/support.fhir-guide.conventions#:code-systems-naming) - назван неправильно, это описание систем идентификации, а названо "Coding Systems"


5. Для описанных профилей не приведены примеры экземпляров, а без них полноценная проверка не возможно.
Необходимо добавить примеры экземпляров, соответствующих обозначенному профилю.
Должен быть приведён хотя бы один пример экземпляра, в котором заполнено максимально возможное количество атрибутов.
В дополнении к нему можно также добавить вспомогательные примеры  - с минимальным заполнением или с заполнением под конкретный сценарий (для лаборатории, например).

**Reply by avzviagin** [reply]
Внес рекомендованные исправления. 
Сделал два профиля: CoreServiceRequest и PoctServiceRequest. Профили без ошибок проходят генерацию (https://fshschool.org/FSHOnline/#/) без примеров. 
Добавил пример CoreServiceRequest с примером. Не проходят кодовые значения.
[CoreServiceRequest.txt](https://github.com/fhir-ru/core/files/11657731/CoreServiceRequest.txt)
[CoreServiceRequest+examlpe.txt](https://github.com/fhir-ru/core/files/11657732/CoreServiceRequest%2Bexamlpe.txt)
[PoctServiceRequest.txt](https://github.com/fhir-ru/core/files/11657733/PoctServiceRequest.txt)

**Reply by Vrodvi** [reply]
1) Переопределять стандартные наборы значений не нужно:
```
* status from http://hl7.org/fhir/ValueSet/request-status (required)
* intent from http://hl7.org/fhir/ValueSet/request-intent (required)
```

2) Профиль для "Группы" должен быть унаследован от ресурса Group (http://hl7.org/fhir/group.html), а у вас указан Patient:
```
Profile:     CoreGroup
Id:          core-group
Parent:      Patient
Title:       "TODO (Ru Core)"
Description: "TODO"
```

3) Кодированные значения в экземплярах ресурсов (в примерах / instance) должны указывать через спецсимвол `#`, а у вас они заданы как строки:
вместо:
```
* status = "completed"
* intent = "original-order" 
```
должно быть:
```
* status = #completed
* intent = #original-order
```

4) Ссылка на пациента дана также в виде текстовой строки из-за чего она никак не может быть провалидирована, а ведь в этом месте требуется не просто ссылка на любого пациента, а именно ссылка на пациента с профилем CorePatient. 
Чтобы осуществлялась валидация профилей по ссылкам нужно использовать синтаксис через `Reference(тут-идентификатор-экземпляра-на-который-идёт-ссылка)`, а также должны существовать экземпляры, на которые идут ссылки и эти экземпляры должны соответствовать указанным профилям.
Вместо:
```
* subject.reference = "Patient/1"
```
должно быть:
```
...
* subject = Reference(core-patient-instance-1)
...

Instance:   core-patient-instance-1
InstanceOf: CorePatient
Usage: #example
```

5) [ОПЦИОНАЛЬНОЕ стилистическое замечание] Для описания кодов в экземплярах ресурсов можно использовать более компактный формат `url-для-codesystem#код "расшифровка кода"`
Т.е. вместо:
```
* code.coding.code = "49541-6"
* code.coding.display = "Fasting status - Reported"
* code.coding.system = "http://loinc.org"
```
можно написать:
```
* code = http://loinc.org#49541-6 "Fasting status - Reported"
```

**Reply by eugenekogan-spb** [reply]
> 4.1. @eugenekogan-spb Тут ещё небольшая проблема в документации на Zendoc - раздел [Идентификаторы Coding Systems](http://fhir-ru.zendoc.me/support.fhir-guide.conventions#:code-systems-naming) - назван неправильно, это описание систем идентификации, а названо "Coding Systems"

поправил: теперь называется [Системы идентификации (Naming Systems)](https://fhir-ru.zendoc.me/support.fhir-guide.conventions#:naming-systems-naming)

**Comment by avzviagin** [comment]
Внес изменения в CoreServiceRequest с примером: генерация без ошибок.
[CoreServiceRequest+examlpe.txt](https://github.com/fhir-ru/core/files/11663321/CoreServiceRequest%2Bexamlpe.txt)
Нужно ли в дочернем профиле повторять описание унаследованных атрибутов? 
Если нет, то как будет при генерации валидироваться пример, указанный в дочернем профиле?

**Reply by Vrodvi** [reply]
Повторять описание унаследованных атрибутов не нужно.
Валидироваться будут все наложенные ограничения по всей иерархии профилей.

**Reply by Vrodvi** [reply]
Пример:
![image](https://github.com/fhir-ru/core/assets/92928137/81c5a344-1ff1-4de1-b325-e0f7c0c116fd)
Вот я уточнил в дочернем профиле только один атрибут, но валидируются и все предыдущие ограничения из родительского профиля.
Я закоментировал в экземпляре указание subject - и на него валидация заругалась, но если разкоментировать, то валидация пройдёт:
![image](https://github.com/fhir-ru/core/assets/92928137/a246664c-a7e0-4713-956c-c991ee39f22c)
